server/src/main/java/org/elasticsearch/action/ActionModule.java

```diff
@@ -844,9 +846,19 @@ public <Request extends ActionRequest, Response extends ActionResponse> void reg
     }
 
     private static ActionFilters setupActionFilters(List<ActionPlugin> actionPlugins) {
-        return new ActionFilters(
-            Collections.unmodifiableSet(actionPlugins.stream().flatMap(p -> p.getActionFilters().stream()).collect(Collectors.toSet()))
-        );
+        List<ActionFilter> finalFilters = new ArrayList<>();
+        List<MappedActionFilter> mappedFilters = new ArrayList<>();
+        for (var plugin : actionPlugins) {
+            for (var filter : plugin.getActionFilters()) {
+                if (filter instanceof MappedActionFilter mappedFilter) {
```

Comment:
nit/idea:
I don't like the cast too much. Would it be difficult to change `plugin.getActionFilters()`?
like 
```
interface ActionFiltersRegistar {
  registerFilter(ActionFilter);
  registerMappedFilter(MappedActionFilter);
}

plugin.getActionFilters(new ActionFiltersRegistar {
   registerMappedFilter(f) { 
      mappedFilters.add(f);
   }
})
```

Reply:
I don't want to complicate the action plugin interface. The above design has an extra level of indirection. It also allows plugins to do funny things like hold onto the registrar and then callback later. What would we expect then?

While I understand the dislike for the cast, I found it necessary in this case because we are trying to fit in within the existing action filters hierarchy. A MappedActionFilter _is_ an ActionFilter, it's just a special case.

Reply:
> funny things like hold onto the registrar and then callback later

Yeah, that would be really bad and require extra care from the implementation (like using internal state to throw if it "survives" beyond the desired scope). 
Fair enough on the cast; if there was a better way of doing it I would have tried it, but if it's necessary no big deal.